<a href="https://colab.research.google.com/github/DangHuuLong/Ai-Recruiter-Mini-Ai-Service/blob/experiment%2Ffine-tune-similarity-colab-v0.2/notebooks/fine_tune_similarity_v0.2_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fine-tune CV-JD Similarity Model v0.2

This notebook is a thin Colab wrapper around `training/fine_tune_similarity.py`.

Use GPU runtime before running the training cells.

## 1. Clone repository and checkout branch

In [3]:
!git clone https://github.com/DangHuuLong/Ai-Recruiter-Mini-Ai-Service.git
%cd Ai-Recruiter-Mini-Ai-Service
!git checkout experiment/fine-tune-similarity-colab-v0.2

fatal: destination path 'Ai-Recruiter-Mini-Ai-Service' already exists and is not an empty directory.
/content/Ai-Recruiter-Mini-Ai-Service
Already on 'experiment/fine-tune-similarity-colab-v0.2'
Your branch is up to date with 'origin/experiment/fine-tune-similarity-colab-v0.2'.


In [4]:
%cd /content/Ai-Recruiter-Mini-Ai-Service
!git status

/content/Ai-Recruiter-Mini-Ai-Service
On branch experiment/fine-tune-similarity-colab-v0.2
Your branch is up to date with 'origin/experiment/fine-tune-similarity-colab-v0.2'.

nothing to commit, working tree clean


In [4]:
!nvidia-smi

Sun Jun  7 08:51:18 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 580.82.07              Driver Version: 580.82.07      CUDA Version: 13.0     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  Tesla T4                       Off |   00000000:00:04.0 Off |                    0 |
| N/A   36C    P8              9W /   70W |       0MiB /  15360MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

## 2. Install dependencies

In [5]:
!pip install -r requirements.txt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 108.8/108.8 kB 10.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 4.2 MB/s eta 0:00:00
INFO: pip is looking at multiple versions of huggingface-hub to determine which version is compatible with other requirements. This could take a while.
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.7/135.7 kB 13.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 110.5/110.5 kB 11.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 117.7/117.7 kB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 517.7/517.7 kB 33.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.6/68.6 kB 6.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 472.0/472.0 kB 38.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.9/60.9 kB 5.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 83.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 375.2

## 3. Validate dataset

In [5]:
!python training/validate_dataset.py --dataset-root datasets --version v0.2
!python training/check_fine_tune_readiness.py --dataset-root datasets --version v0.2

Dataset validation passed.
Pair dataset: datasets/versions/v0.2/cv_jd_pairs.jsonl
Total pairs: 2275

Fine-tuning readiness check passed.
Next action: create a separate fine-tuning experiment branch and keep generated artifacts out of git.


## 4. Smoke test

This checks that the script, dataset, and model dependencies work. Do not use this result for model comparison.

In [6]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/debug-miniLM-v0.2 \
  --report-path artifacts/reports/debug_fine_tune_similarity_v0.2_report.json \
  --epochs 1 \
  --batch-size 2 \
  --max-train-samples 20 \
  --max-eval-samples 20

2026-06-07 08:53:35.474326: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
modules.json: 100% 349/349 [00:00<00:00, 2.12MB/s]
config_sentence_transformers.json: 100% 116/116 [00:00<00:00, 962kB/s]
README.md: 10.5kB [00:00, 31.7MB/s]
sentence_bert_config.json: 100% 53.0/53.0 [00:00<00:00, 454kB/s]
config.json: 100% 612/612 [00:00<00:00, 4.61MB/s]
model.safetensors: 100% 90.9M/90.9M [00:01<00:00, 83.9MB/s]
tokenizer_config.json: 100% 350/350 [00:00<00:00, 2.11MB/s]
vocab.txt: 232kB [00:00, 29.7MB/s]
tokenizer.json: 466kB [00:00, 45.9MB/s]
special_tokens_map.json: 100% 112/112 [00:00<00:00, 821kB/s]
config.json: 100% 190/190 [00:00<00:00, 1.53MB/s]
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniL

## 5. Full fine-tuning run

In [7]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_report.json \
  --epochs 1 \
  --batch-size 8

2026-06-07 09:04:37.932850: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 1
Batch size: 8
Warmup steps: 20
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 6. Review report

In [8]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 19.3292,
  'rmse': 24.5166,
  'label_accuracy': 0.2769,
  'mean_similarity': 0.59864,
  'mean_predicted_score': 79.9262,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 19.8629,
  'rmse': 24.7948,
  'label_accuracy': 0.3057,
  'mean_similarity': 0.586054,
  'mean_predicted_score': 79.2857,
  'mean_target_score': 62.3886}}

## 7. Save artifacts to Google Drive


In [9]:
from google.colab import drive
drive.mount('/content/drive')

import shutil, pathlib
dest = pathlib.Path('/content/drive/MyDrive/ai-recruiter-artifacts')
dest.mkdir(parents=True, exist_ok=True)

shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2',
    dest / 'fine-tuned-miniLM-v0.2',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_report.json',
    dest / 'fine_tune_similarity_v0.2_report.json'
)
print("Saved to Drive")


Mounted at /content/drive
Saved to Drive


## 8. Experiment 2 - epochs=2

Epoch 1 did not beat baseline on MAE/RMSE. Trying 2 epochs.

Baseline to beat: MAE < 17.6431 | RMSE < 22.4497 | Label accuracy > 0.2822


In [10]:
!python training/fine_tune_similarity.py \
  --dataset-root datasets \
  --version v0.2 \
  --base-model sentence-transformers/all-MiniLM-L6-v2 \
  --output-dir artifacts/models/fine-tuned-miniLM-v0.2-ep2 \
  --report-path artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json \
  --epochs 2 \
  --batch-size 8


2026-06-07 09:29:18.964241: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Fine-tuning CV-JD similarity model.
Dataset version: v0.2
Base model: sentence-transformers/all-MiniLM-L6-v2
Train examples: 1600
Validation examples: 325
Test examples: 350
Epochs: 2
Batch size: 8
Warmup steps: 40
wandb: (1) Create a W&B account
wandb: (2) Use an existing W&B account
wandb: (3) Don't visualize my results
wandb: Enter your choice: 3
wandb: You chose "Don't visualize my results"
wandb: Using W&B in offline mode.
wandb: W&B API key is configured. Use `wandb login --relogin` to force relogin
wandb: Tracking run with wandb version 0.27.0
wandb: W&B syncing is set to `offline` in this directory. Run `wandb online` or set WANDB_MODE=online to enable cloud synci

## 9. Review report - experiment 2


In [11]:
import json
from pathlib import Path

report_path = Path('artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json')
report = json.loads(report_path.read_text(encoding='utf-8'))
report['metrics']


{'validation': {'pair_count': 325,
  'mae': 19.8892,
  'rmse': 25.033,
  'label_accuracy': 0.2738,
  'mean_similarity': 0.616088,
  'mean_predicted_score': 80.7938,
  'mean_target_score': 62.6892},
 'test': {'pair_count': 350,
  'mae': 20.5371,
  'rmse': 25.7064,
  'label_accuracy': 0.2971,
  'mean_similarity': 0.609783,
  'mean_predicted_score': 80.48,
  'mean_target_score': 62.3886}}

## 10. Save artifacts - Experiment 2 (epochs=2) to Google Drive

Epoch 2 performed worse than epoch 1 across all metrics. Saving artifacts for record only.

| Metric       | Baseline | Epoch 1 | Epoch 2 |
|--------------|----------|---------|---------|
| Val MAE      | 17.6431  | 19.3292 | 19.8892 |
| Val RMSE     | 22.4497  | 24.5166 | 25.033  |
| Val Acc      | 0.2822   | 0.2769  | 0.2738  |
| Test Acc     | 0.2822   | 0.3057  | 0.2971  |

Neither run beats the baseline. More epochs made results worse — sign of wrong training direction.
Do not integrate this model into service inference.


In [12]:
shutil.copytree(
    'artifacts/models/fine-tuned-miniLM-v0.2-ep2',
    dest / 'fine-tuned-miniLM-v0.2-ep2',
    dirs_exist_ok=True
)
shutil.copy(
    'artifacts/reports/fine_tune_similarity_v0.2_ep2_report.json',
    dest / 'fine_tune_similarity_v0.2_ep2_report.json'
)


PosixPath('/content/drive/MyDrive/ai-recruiter-artifacts/fine_tune_similarity_v0.2_ep2_report.json')